{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "\"\"\"Contains function to perform GIM3E\"\"\"\n",
    "\n",
    "from __future__ import absolute_import\n",
    "\n",
    "import six\n",
    "from optlang.symbolics import Zero\n",
    "from cobra.util import fix_objective_as_constraint\n",
    "\n",
    "def gim3e(model, expression_profile, fraction_of_optimum, condition=None):\n",
    "    \"\"\"\n",
    "    Perform GIM3E (Gene Inactivation Moderated by Metabolism, Metabolomics and Expression)\n",
    "    on a GEM, to enable the integrated functional analysis of intracellular metabolomics\n",
    "    data and gene expression microarray data. The models created with the GIM3E algorithm\n",
    "    report the modeled rate of creation or consumption (turnover) of metabolites.[1]\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    model: cobra.Model\n",
    "        A constraint-based model to perform GIMME on.\n",
    "    expression_profile: ExpressionProfile\n",
    "        An expression profile to integrate in the model.\n",
    "    fraction_of_optimum: float\n",
    "        The fraction of the Required Metabolic Functionalities.\n",
    "    condition: str or int, optional\n",
    "        The condition from the expression profile. If None, the first condition is used.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    model: cobra.Model\n",
    "    sol.objective_value: float\n",
    "\n",
    "    References:\n",
    "    ----------\n",
    "    .. [1] Brian J. Schmidt, Ali Ebrahim, Thomas O. Metz, Joshua N. Adkins, Bernhard Ø. Palsson, Daniel R. Hyduke;\n",
    "           GIM3E: condition-specific models of cellular metabolism developed from metabolomics and expression data,\n",
    "           Bioinformatics, Volume 29, Issue 22, 15 November 2013, Pages 2900–2908,\n",
    "           https://doi.org/10.1093/bioinformatics/btt493\n",
    "    \"\"\"\n",
    "    assert isinstance(model, cobra.Model)\n",
    "    assert isinstance(expression_profile, ExpressionProfile)\n",
    "    assert isinstance(fraction_of_optimum, float)\n",
    "    assert isinstance(condition, (str, int, None))\n",
    "    \n",
    "    with model:\n",
    "        prob = model.problem\n",
    "        condition = expression_profile.conditions[0] if condition is None else condition\n",
    "        reaction_profile = expression_profile.to_reaction_dict(condition, model, not_measured_value, normalization)\n",
    "        # Step 1: FBA\n",
    "        model.objective.direction = \"max\"\n",
    "        o_opt = model.slim_optimize()\n",
    "        fix_objective_as_constraint(model, fraction=fraction_of_optimum)\n",
    "        # Step 2: Add turnover metabolites\n",
    "        # 2.1 Make model irreversible\n",
    "        for rxn in model.reactions:\n",
    "            if rxn.reversibility is True:\n",
    "                reverse_rxn = rxn.copy()\n",
    "                # Fix names\n",
    "                reverse_rxn.name = rxn.name + ' reverse'\n",
    "                rxn.name += ' forward'\n",
    "                # Fix IDs\n",
    "                reverse_rxn.id = rxn.id + '_reverse'\n",
    "                rxn.id += '_forward'\n",
    "                # Fix bounds\n",
    "                reverse_rxn.bounds = (0, -rxn.lower_bound)\n",
    "                rxn.bounds = (0, rxn.upper_bound)\n",
    "                # Fix metabolite coefficients\n",
    "                reverse_rxn.subtract_metabolites(reverse_rxn.metabolites, combine=False)\n",
    "                # Add the reverse reaction to the model\n",
    "                model.add_reaction(reverse_rxn)\n",
    "        # 2.2 Turnover metabolites\n",
    "        turnover_mets = []\n",
    "        for met in model.metabolites:\n",
    "            turnover_met = met.copy()\n",
    "            turnover_met.name = met.name + ' turnover'\n",
    "            turnover_met.id = met.id + '_turnover'\n",
    "            turnover_mets.append(turnover_met)\n",
    "        irr_model.add_metabolites(turnover_mets)\n",
    "        # 2.3 Add turnover metabolites to reactions\n",
    "        for rxn in model.reactions:\n",
    "            mets_to_add = {}\n",
    "            for key, value in six.iteritems(rxn.metabolites):\n",
    "                turnover_met = model.metabolites.get_by_id(key.id + '_turnover')\n",
    "                mets_to_add[turnover_met] = abs(value)\n",
    "            rxn.add_metabolites(mets_to_add)\n",
    "        # 2.4 Sink reactions\n",
    "        for met in irr_model.metabolites.query('_turnover$'):\n",
    "            sink_rxn = cobra.Reaction('TMS_' + met.id)\n",
    "            # Since both creation and consumption of\n",
    "            # the real metabolite drive the turnover,\n",
    "            # we require 2 units of metabolite per\n",
    "            # 1 unit of flux sink so this matches\n",
    "            # the turnover through the real metabolite.\n",
    "            sink_rxn.add_metabolites({met: -2})\n",
    "            sink_rxn.lower_bound = 1.01 * model.solver.configuration.tolerances.feasibility        \n",
    "            irr_model.add_reaction(sink_rxn)\n",
    "        # Step 3: Add penalty coefficients\n",
    "        penalty_coefficients = {r: i_max - i_g for r, exp in six.iteritems(reaction_profile)}\n",
    "        # Step 4: Determine penalty bounds\n",
    "        obj_vars = []\n",
    "        for rxn_id, coefficient in six.iteritems(coefficients):\n",
    "            rxn = model.reactions.get_by_id(rxn_id)\n",
    "            obj_vars.append(coefficient * rxn.flux_expression)\n",
    "\n",
    "        model.objective = prob.Objective(Zero, sloppy=True, direction=\"min\")\n",
    "        model.objective.set_linear_coefficients({v: 1.0 for v in obj_vars})\n",
    "            \n",
    "        sol = model.optimize()\n",
    "        return model"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py35-cobra-cplex",
   "language": "python",
   "name": "py35-cobra-cplex"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
